kernel/arch/dreamcast/sound/snd_stream.c

```diff
@@ -151,16 +151,40 @@ static void process_filters(snd_stream_hnd_t hnd, void **buffer, int *samplecnt)
 }
 
 static void sep_data(void *buffer, int len, int stereo) {
-    uint32_t data;
     uint32_t *buf = (uint32_t *)buffer;
-    uint16_t *left = (uint16_t *)sep_buffer[1];
-    uint16_t *right = (uint16_t *)sep_buffer[0];
+    uint32_t *left_ptr = (uint32_t *)sep_buffer[1];
+    uint32_t *right_ptr = (uint32_t *)sep_buffer[0];
+    uint32_t data;
+    uint32_t left_val;
+    uint32_t right_val;
 
     if(stereo) {
-        for(; len > 0; len -= 2) {
+        len <<= 1;
+        for(; len > 8; len -= 8) {
+            if(((uintptr_t)buf & 31) == 0) {
+                dcache_pref_block(buf + 8);
+            }
+            data = *buf++;
+            left_val = (data >> 16);
+            right_val = (data & 0xffff);
+
+            data = *buf++;
+            left_val |= (data & 0xffff0000);
+            right_val |= (data & 0xffff) << 16;
```

Comment:
It might be possible to make this loop faster by reading a `uint64_t` and shifting its bits around. Even though the SH-4 is 32-bit, it would give more insight to the compiler about what you're trying to do, which would make it generate better code (using the swap.w + xtrct instructions for instance).

Reply:
Nope, I have some experience with processing 64-bit in ASM, no perfomance profit.
Only FPU can 64-bit, but FPU doesn't have bitwise operations. So need transfer to CPU, process it and then transfer back to FPU by using 32-bit fpul register. So it's make no sense.